minerl/herobraine/env_specs/basalt_specs.py

```diff
@@ -0,0 +1,263 @@
+        return obs
+
+    def step(self, action: dict):
+        observation, reward, done, info = super().step(action)
+        if self._steps_till_done is None:  # Snowball throw hasn't yet been detected.
+            if (observation["equipped_items"]["mainhand"]["type"] == "snowball"
+                    and action["use"] == 1):
+                self._steps_till_done = self.episode_end_delay
+        else:  # Snowball throw was detected. We will end episode soon.
+            if self._steps_till_done == 0:
+                done = True
+                self._steps_till_done = None
+            else:
+                self._steps_till_done -= 1  # pytype: disable=unsupported-operands
+        return observation, reward, done, info
+
+
+def _basalt_gym_entrypoint(
+        env_spec: "BasaltBaseEnvSpec",
+        fake: bool = False,
+        end_after_snowball_throw: bool = True,
+) -> _singleagent._SingleAgentEnv:
+    """Used as entrypoint for `gym.make`."""
+    if fake:
+        env = _fake._FakeSingleAgentEnv(env_spec=env_spec)
+    else:
+        env = _singleagent._SingleAgentEnv(env_spec=env_spec)
+    if end_after_snowball_throw:
+        env = EndAfterSnowballThrowWrapper(env)
+    return env
+
+
+BASALT_GYM_ENTRY_POINT = "minerl.herobraine.env_specs.basalt_specs:_basalt_gym_entrypoint"
+
+
+class BasaltBaseEnvSpec(EnvSpec):
+
+    def __init__(
+            self,
+            name,
+            demo_server_experiment_name,
+            max_episode_steps=2400,
+    ):
+        assert "/" not in demo_server_experiment_name
+        super().__init__(name=name, max_episode_steps=max_episode_steps)
+        self.demo_server_experiment_name = demo_server_experiment_name
+
+    def is_from_folder(self, folder: str) -> bool:
+        # Implements abstractmethod.
+        return folder == self.demo_server_experiment_name
+
+    def _entry_point(self, fake: bool) -> str:
+        # Assuming here that `fake` argument gets passed along to entrypoint fn.
+        return BASALT_GYM_ENTRY_POINT
+
+    def create_observables(self):
+        return DEFAULT_OBS_HANDLERS
+
+    def create_actionables(self):
+        return DEFAULT_ACT_HANDLERS
+
+    def create_agent_handlers(self) -> List[handlers.Handler]:
+        return []
+
+    def create_server_quit_producers(self) -> List[handlers.Handler]:
+        return [
+            handlers.ServerQuitFromTimeUp(
+                (self.max_episode_steps * mc.MS_PER_STEP)),
+            handlers.ServerQuitWhenAnyAgentFinishes()
+        ]
+
+    def create_server_decorators(self) -> List[handlers.Handler]:
+        return []
+
+    def create_server_initial_conditions(self) -> List[handlers.Handler]:
+        return [
+            handlers.TimeInitialCondition(
+                allow_passage_of_time=False
+            ),
+            handlers.SpawningInitialCondition(
+                allow_spawning=True
+            )
+        ]
+
+    def auto_blacklist(self, npz_data: dict) -> Optional[str]:
```

Comment:
Adding a TODO for later to make auto_blacklist a part of the envspec as an optional override